flang/lib/Lower/Allocatable.cpp

```diff
@@ -42,12 +42,74 @@ static llvm::cl::opt<bool> useDescForMutableBox(
     llvm::cl::desc("Always use descriptors for POINTER and ALLOCATABLE"),
     llvm::cl::init(false));
 
+//===----------------------------------------------------------------------===//
+// Error management
+//===----------------------------------------------------------------------===//
+
+namespace {
+// Manage STAT and ERRMSG specifier information across a sequence of runtime
+// calls for an ALLOCATE/DEALLOCATE stmt.
+struct ErrorManager {
+  void init(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+            const Fortran::lower::SomeExpr *statExpr,
+            const Fortran::lower::SomeExpr *errMsgExpr) {
+    Fortran::lower::StatementContext stmtCtx;
+    builder = &converter.getFirOpBuilder();
+    hasStat = builder->createBool(loc, statExpr != nullptr);
+    statAddr = statExpr
+                   ? fir::getBase(converter.genExprAddr(statExpr, stmtCtx, loc))
+                   : mlir::Value{};
+    if (statExpr && errMsgExpr) {
+      auto errMsgDesc = builder->createBox(
+          loc, converter.genExprAddr(errMsgExpr, stmtCtx, loc));
+      errMsgAddr = builder->createTemporary(loc, errMsgDesc.getType());
+      builder->create<fir::StoreOp>(loc, errMsgDesc, errMsgAddr);
+    } else {
+      errMsgAddr = builder->createNullConstant(loc);
+    }
+    sourceFile = Fortran::lower::locationToFilename(*builder, loc);
+    sourceLine = Fortran::lower::locationToLineNo(*builder, loc,
+                                                  builder->getIntegerType(32));
+  }
+
+  bool hasStatSpec() const { return static_cast<bool>(statAddr); }
+
+  void genStatCheck(mlir::Value stat, mlir::Location loc) {
+    if (hasStatSpec() && stat) {
+      auto zero = builder->createIntegerConstant(
+          loc, fir::dyn_cast_ptrEleTy(statAddr.getType()), 0);
+      auto cmp = builder->create<mlir::CmpIOp>(loc, mlir::CmpIPredicate::eq,
+                                               stat, zero);
+      auto ifOp = builder->create<fir::IfOp>(loc, cmp,
+                                             /*withElseRegion=*/false);
+      builder->setInsertionPointToStart(&ifOp.thenRegion().front());
+    }
+  }
+
+  void assignStat(mlir::Value stat, mlir::Location loc) {
+    if (hasStatSpec()) {
+      assert(stat && "missing stat value");
+      auto castStat = builder->createConvert(
+          loc, fir::dyn_cast_ptrEleTy(statAddr.getType()), stat);
+      builder->create<fir::StoreOp>(loc, castStat, statAddr);
+    }
+  }
+
+  Fortran::lower::FirOpBuilder *builder;
```

Comment:
I  would avoid saving the builder here unless there is a functional point of doing so, I understand it saves and argument but it brings a bit more complexity given the ErrorManager life time is quite long.

We have a flow doing:
```
errorManager.init(converter); // capture pointer to builder
... use builder/potentially change something about it
errorManager.doSomethingWithTheBuilderFromInit()
```

I get that functionally, nothing weird is happening to the builder between the calls, we are not creating a new one, but I think makes code audit more complex.

Reply:
Done